src/journal/bsod.c

```diff
@@ -0,0 +1,195 @@
+#include "fd-util.h"
+#include "io-util.h"
+#include "log.h"
+#include "main-func.h"
+#include "sd-id128.h"
+#include "sd-journal.h"
+#include "terminal-util.h"
+
+static void help(void) {
+        printf("%s\n\n"
+               "filters the journal to fetch the first message from the current boot with an emergency log level.\n\n"
+               "   -h --help            Show this help\n"
+               "      --version         Show package version\n",
+               program_invocation_short_name);
+}
+
+static int acquire_first_emergency_log_message(char **ret) {
+        _cleanup_(sd_journal_closep) sd_journal *j = NULL;
+        const void *d;
+        size_t l;
+        sd_id128_t boot_id;
+        char boot_id_filter[42];
+        int r;
+
+        r = sd_journal_open(&j, SD_JOURNAL_LOCAL_ONLY);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open journal: %m");
+
+        r = sd_id128_get_boot(&boot_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get boot ID: %m");
+
+        snprintf(boot_id_filter, sizeof(boot_id_filter), "_BOOT_ID=" SD_ID128_FORMAT_STR, SD_ID128_FORMAT_VAL(boot_id));
+
+        r = sd_journal_add_match(j, boot_id_filter, 0);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to add boot ID filter: %m");
+
+        r = sd_journal_add_match(j, "_UID=0", 0);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to add User ID filter: %m");
+
+        assert_cc(0 == LOG_EMERG);
+        r = sd_journal_add_match(j, "PRIORITY=0", 0);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to add Emergency filter: %m");
+
+        r = sd_journal_seek_head(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to seek to start of jornal: %m");
+
+        r = sd_journal_next(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read next journal entry: %m");
+        if (r == 0) {
+                log_warning("No entries in the journal");
+                return -ENOENT;
+        }
+
+        r = sd_journal_get_data(j, "MESSAGE", &d, &l);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read journal message: %m");
+
+        *ret = strndup((char *) d, l);
+        if (! *ret) {
+                r = -ENOMEM;
+                return log_error_errno(r, "Failed to duplicate journal message: %m");
+        }
+
+        return r;
+}
+
+static int find_next_free_vt(int fd) {
+        size_t i;
+        struct vt_stat terminal_status;
+
+        if (ioctl(fd, VT_GETSTATE, &terminal_status) < 0)
+                return -errno;
+
+        for (i = 0; i < sizeof(terminal_status.v_state) * 8; i++) {
+                if ((terminal_status.v_state & (1 << i)) == 0)
+                        break;
+        }
+        return i;
+}
+
+static int display_emergency_message_fullscreen(char * message) {
+        assert(message);
+        int r;
+        char tty[20];
```

Comment:
use `char tty[STRLEN("/dev/tty") + DECIMAL_STR_MAX(int) + 1]` to size the string properly